lisa/runner.py

```diff
@@ -282,52 +291,41 @@ def _start_loop(self) -> None:
                     f"{task_manager._futures}"
                 )
 
-                for runner in remaining_runners:
-                    while not runner.is_done:
-                        # fetch a task and submit
-                        task = runner.fetch_task()
-                        if task:
-                            # the message may be too long to display.
-                            task_message = str(task)
-                            task_message = (
-                                task_message
-                                if len(task_message) < 200
-                                else f"{task_message[:200]}..."
-                            )
+                while task_manager.has_idle_worker():
+                    for runner in remaining_runners:
+                        while not runner.is_done and task_manager.has_idle_worker():
+                            # fetch a task and submit
+                            task = runner.fetch_task()
+                            if task:
+                                task_manager.submit_task(task)
+                            else:
+                                # current runner may not be done, but it doesn't
+                                # have task temporarily. The root runner can start
+                                # tasks from next runner.
+                                self._log.debug(
+                                    f"No task available for runner: {runner.id}"
+                                )
+                                break
+                        if runner.is_done:
+                            # remove fully completed runner.
+                            runner.close()
+                            remaining_runners.remove(runner)
                             self._log.debug(
-                                f"fetched task from {runner.id}: '{task_message}'"
+                                f"runner '{runner.id}' is done, "
```

Comment:
Because there is a line of log in `runner.close()` already, this line can be removed. In practice, the list of remaining runner doesn't help a lot.